python/ray/data/_internal/datasource/lance_datasink.py

```diff
@@ -0,0 +1,266 @@
+from itertools import chain
+from typing import (
+    Any,
+    Dict,
+    Literal,
+    Optional,
+    TYPE_CHECKING,
+    List,
+    Tuple,
+    Iterable,
+    Union,
+)
+import pickle
+
+from ray.data._internal.util import _check_import
+from ray.data.datasource.datasink import Datasink
+import pyarrow as pa
+
+
+if TYPE_CHECKING:
+    import pandas as pd
+    from lance.fragment import FragmentMetadata
+
+
+def _pd_to_arrow(
+    df: Union["pa.Table", "pd.DataFrame", Dict], schema: Optional[pa.Schema]
+) -> "pa.Table":
+    """Convert a pandas DataFrame to pyarrow Table."""
+    import pandas as pd
+
+    if isinstance(df, dict):
+        return pa.Table.from_pydict(df, schema=schema)
+    elif isinstance(df, pd.DataFrame):
+        tbl = pa.Table.from_pandas(df, schema=schema)
+        tbl = tbl.replace_schema_metadata()
+        return tbl
+    elif isinstance(df, pa.Table):
+        if schema is not None:
+            return df.cast(schema)
+    return df
+
+
+def _write_fragment(
+    stream: Iterable[Union["pa.Table", "pd.DataFrame"]],
+    uri: str,
+    *,
+    schema: Optional["pa.Schema"] = None,
+    max_rows_per_file: int = 1024 * 1024,
+    max_bytes_per_file: Optional[int] = None,
+    max_rows_per_group: int = 1024,  # Only useful for v1 writer.
+    data_storage_version: Optional[str] = None,
+    storage_options: Optional[Dict[str, Any]] = None,
+) -> List[Tuple["FragmentMetadata", "pa.Schema"]]:
+    import pandas as pd
+    from lance.fragment import write_fragments, DEFAULT_MAX_BYTES_PER_FILE
+
+    if schema is None:
+        first = next(stream)
+        if isinstance(first, pd.DataFrame):
+            schema = pa.Schema.from_pandas(first).remove_metadata()
+        elif isinstance(first, Dict):
+            tbl = pa.Table.from_pydict(first)
+            schema = tbl.schema.remove_metadata()
+        else:
+            schema = first.schema
+        if len(schema.names) == 0:
+            # Empty table.
+            schema = None
+
+        stream = chain([first], stream)
+
+    def record_batch_converter():
+        for block in stream:
+            tbl = _pd_to_arrow(block, schema)
+            yield from tbl.to_batches()
```

Comment:
There are already a util function for converting data format. 
You can use do
```python
BlockAccessor.for_block(block).to_arrow()
```

Reply:
Done